Provide a summary view of recent diaries

nertc opened the pull request:
This PR addresses "Provide a summary view of recent diaries" issue mentioned in the https://github.com/openstreetmap/openstreetmap-website/issues/3887

[Bootstrap Accordion](https://getbootstrap.com/docs/5.2/components/accordion/) was added to the diary entries list pages. Now the content of the diary entry will be collapsed until toggle button is clicked. View of the diary entry details (show) page remains the same. Fixes https://github.com/openstreetmap/openstreetmap-website/issues/3887

Screenshots:
![Screenshot 2024-08-26 122446](https://github.com/user-attachments/assets/aaefa241-9f48-4474-8061-73e22d40ae3e)
![Screenshot 2024-08-26 122459](https://github.com/user-attachments/assets/32076fb2-f92d-46c2-8e7c-170b204cb6bc)
![Screenshot 2024-08-26 122517](https://github.com/user-attachments/assets/29b902d9-8460-4ba4-a4a1-098e566d057c)

nertc commented:
Failed test is out of scope of this PR, as there was no change or impact related to the test that was failed.

nertc commented:
This PR also partially refers to https://github.com/openstreetmap/openstreetmap-website/issues/3721 (changes HTML view of the diary entries list, but doesn't change RSS)

mmd-osm commented:
I haven't tested this code yet, though it seems that toggling a diary entry would either show the whole text, or only the title. I believe #3887 talked about showing the first few _n_ lines of the body, with an option to further expand the body of the diary. Would this be possible to implement?

tomhughes commented:
Yes this starts with all entries collapsed to just their headings.

It also fails if you have CSP enforcement enabled because bootstrap is using `data:` urls for the arrow images so they don't appear unless you allow that in the controller with:

```ruby
content_security_policy(:only => :index) do |policy|
  policy.img_src(*policy.img_src, "data:")
end
```

nertc commented:
@mmd-osm Yes, it is possible. I made this change according to the last screen mentioned in the Issue https://github.com/openstreetmap/openstreetmap-website/issues/3887. I'll change this PR to show part of the content even when it is collapsed.

@tomhughes I'll take a look if there is any way to make it work without custom code.

nertc commented:
@mmd-osm I updated PR so that now `10 rem` (about 7.5 paragraph lines) of content is shown when the entry is collapsed. If entry is shorter, it doesn't leave any redundant gaps. I needed to add 2 lines of custom CSS as Bootstrap Accordion doesn't support behavior of showing any content when collapsed (it sets `display: none`). The only downside of the current solution is that when entry is less than 7 lines, accordion button is still shown but clicking it doesn't change anything. The only fix (based on limited functionality of Bootstrap Accordion) can be made by adding custom JS code. From my experience, we try to avoid adding additional JS code. If it is okay to add one for this functionality, I can add it too.

@tomhughes I implemented solution suggested by you. I found some issues opened about Bootstrap CSP errors (https://github.com/twbs/bootstrap/issues/25394), but even [Bootstrap 5 mentions](https://getbootstrap.com/docs/5.0/customize/overview/#csps-and-embedded-svgs) that some of their functionality uses `data:` in image urls. As mentioned in [Stackexchange](https://security.stackexchange.com/questions/94993/is-including-the-data-scheme-in-your-content-security-policy-safe/167244#167244) and [MDN](https://developer.mozilla.org/en-US/docs/Web/SVG/SVG_as_an_Image#restrictions) adding `data:` as `background-image` or other image source doesn't raise any risks. Therefore, if we don't have extremely strict CSP policy, we can use solution suggested by you.

Thank you for reviewing

tomhughes commented:
I guess the question is do we really want/need the collapse and expand functionality or is it better just to show a truncated post and let people click through if they want to see the whole thing, which I think is a more typical way to handle this sort of thing.

nertc commented:
I created a separate PR for the truncated version of the entry https://github.com/openstreetmap/openstreetmap-website/pull/5121. If we decide to truncate diary entries instead of showing summary of them, I'll close this PR.

nertc commented:
Closing this PR in favor of #5121 PR
